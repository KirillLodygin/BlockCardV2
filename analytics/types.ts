import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';

import { LOCK_OPERATION } from '../types';
import {
  CREDIT_ACTION_PERMANENT_BLOCK_STRING,
  CREDIT_ACTION_TEMPORARY_BLOCK_STRING,
  CREDIT_CARD_ACTION_MAP,
  CREDIT_PRODUCT,
  CREDIT_STREAM,
  DEBIT_ACTION_PERMANENT_BLOCK_STRING,
  DEBIT_ACTION_TEMPORARY_BLOCK_STRING,
  DEBIT_CARD_ACTION_MAP,
  DEBIT_PRODUCT,
  DEBIT_STREAM,
} from './constants';

export type AnalyticAction = {
  text: string;
  target: string;
};

// для аналитики
export enum CARD_ACTION_TYPE {
  UNFREEZE = 'unfreeze',
  BLOCK = 'block',
}

export type AnalyticsMapType = {
  [x in string]: {
    [x in string]: {
      text: string;
      target: string;
    };
  };
};

export type AnalyticsEntityType = {
  stream: typeof CREDIT_STREAM | typeof DEBIT_STREAM;
  product: typeof CREDIT_PRODUCT | typeof DEBIT_PRODUCT;
  getActionEntityString: (
    blockType: LOCK_OPERATION,
  ) =>
    | typeof DEBIT_ACTION_TEMPORARY_BLOCK_STRING
    | typeof DEBIT_ACTION_PERMANENT_BLOCK_STRING
    | typeof CREDIT_ACTION_TEMPORARY_BLOCK_STRING
    | typeof CREDIT_ACTION_PERMANENT_BLOCK_STRING;
  analyticsMap: AnalyticsMapType;
  type: typeof CARD_ACTION_TYPE.BLOCK | typeof CARD_ACTION_TYPE.UNFREEZE;
};

enum BLOCK_DEBIT_CARD_SCREEN {
  CHECK = 'CHECK',
  RESULT = 'RESULT',
  START = 'START',
}

export type NotificationType = typeof NOTIFICATION_TYPE.info | typeof NOTIFICATION_TYPE.fail;

export type AnalyticsOptionsType = {
  blockType: LOCK_OPERATION;
  cardName: string;
  screenData: BLOCK_DEBIT_CARD_SCREEN;
  notificationType?: NotificationType;
  reason?: string;
};

export type DebitAnalyticsMapType = typeof DEBIT_CARD_ACTION_MAP;
export type CreditAnalyticsMapType = typeof CREDIT_CARD_ACTION_MAP;

export type NotificationKeys = keyof typeof DEBIT_CARD_ACTION_MAP;
