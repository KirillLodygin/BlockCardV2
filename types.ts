import { DropdownValue } from '@vtb/ui-kit/adaptive/molecules/dropdown/_baseDropdown/typesDropdown';

import { ServiceListMsa } from '@modules/CreditCard';
import { EntityByType } from '@modules/DebitAndCreditCards';

export type BlockCardReqeustDataType = {
  secureCode: string;
  transactionId: string;
};

export enum SCENARIO_ASYNC_EVENT_BLOCK_CARD {
  AUTHORIZE = 'card-lock-authorize',
  AUTHORIZE_UNFREEZE = 'card-unfreeze-authorize',
  COMMIT = 'card-lock-commit',
  COMMIT_UNFREEZE = 'card-unfreeze-commit',
  FREEZE = 'card-freeze',
  UNFREEZE = 'card-unfreeze',
}

export type AsyncEventBlockDebitCardResponse = {
  body: AuthorizeResponseType | BlockCardResponseType;
  status: number;
  type: SCENARIO_ASYNC_EVENT_BLOCK_CARD;
};

export type BlockCardResponseType = {
  maskedCardNumber: string;
};

export type AuthorizeResponseType = {
  expirationTime: number;
  phone: string;
  transactionId: string;
};

export type RowTableData = {
  value: string | DropdownValue | null;
  title: string;
};

export enum BLOCK_DEBIT_CARD_SCREEN {
  CHECK = 'CHECK',
  RESULT = 'RESULT',
  START = 'START',
}

export enum BLOCK_TABLE_DATA_TITLE {
  BLOCK_REASON = 'blockReason',
  CARD_TITLE = 'blockCardTitle',
  OPERATION_TYPE = 'operationType',
  SHROT_CRAD_NEMBER = 'shortNumber',
}

export enum LOCK_OPERATION {
  PERMANENT = 'permanent',
  TEMPORARY = 'temporary',
  UNFREEZE = 'unfreeze',
}

export enum BUTTONS_LOCK {
  BACK = 'back',
  CONTINUE = 'continue',
  RESEND = 'resend',
  SECONDARY = 'secondary',
}

export enum NOTIFICATION_TYPE {
  AUTH_ERROR = 'authError',
  ERROR = 'error',
  FAIL = 'fail',
  INFO = 'info',
  SUCCESS = 'success',
}

export enum CARD_TYPE {
  DIGITAL = 'digital',
  PLASTIC = 'plastic',
}

export enum LOCK_REASON {
  LOSS = 'loss',
  THEFT = 'theft',
  FRAUD = 'fraud',
  STUFFY = 'stuffy',
  OTHER = 'other',
}

export type ButtonsLock = {
  isFail?: boolean;
  title: string;
  type: BUTTONS_LOCK;
};

export type NotificationLockResult = {
  description: string;
  operations: LOCK_OPERATION[];
  title?: string;
  type: NOTIFICATION_TYPE;
};

export type LockTypeSelection = {
  buttons: ButtonsLock[];
  buttonsRadio: {
    id: LOCK_OPERATION;
    text: string;
  }[];
  dropdown: {
    elements: {
      id: LOCK_REASON;
      cardTypes: CARD_TYPE[];
      text: string;
    }[];
    label: string;
  };
  inputLabel: string;
  notifications: NotificationLockResult[];
  title: string;
};

export type LockConfirmationScreen = {
  buttons: ButtonsLock[];
  inputErrorText: string;
  inputLabel: string;
  operationName: string;
  phoneInfo: string;
  timerInfo: string;
  table: {
    title: string;
    type: BLOCK_TABLE_DATA_TITLE;
  }[];
  title: string;
};

export type LockResultScreen = {
  buttons: ButtonsLock[];
  notifications: NotificationLockResult[];
};

export type ContentCardBlock = {
  lockTypeSelectionScreen: LockTypeSelection;
  lockConfirmationScreen: LockConfirmationScreen;
  lockResultScreen: LockResultScreen;
};

export type ContentCreditCardBlock = {
  BlockCardHeader?: string;
  BlockCardTemp?: string;
  BlockCardConst?: string;
  BlockCardButton?: string;
  BlockCardDescription?: string;
  TempBlockCardDescription?: string;
  BlockCardConfirmButton?: string;
  BlockCardConfirmHeader?: string;
  BlockCardConfirmTitle?: string;
  BlockCardConfirmNumber?: string;
  BlockCardConfirmType?: string;
  BlockCardConfirmReason?: string;
  SuccessBlockCardHeader?: string;
  SuccessBlockCardDescription?: string;
  SuccessButton?: string;
  ErrorBlockCardHeader?: string;
  ErrorBlockCardDescription?: string;
  ErrorButton?: string;
  NetworkErrorHeader?: string;
  NetworkErrorDescription?: string;
  NetworkErrorAgainButton?: string;
  TempBlockCardHeader?: string;
  SuccessTempBlockCardDescription?: string;
  UnblockCardHeader?: string;
  UnblockCardTitle?: string;
  UnblockCardNumber?: string;
  UnblockCardButton?: string;
  SuccessUnblockCardHeader?: string;
  SuccessUnblockCardDescription?: string;
  ErrorUnblockCardHeader?: string;
  ErrorUnblockCardDescription?: string;
  BlockUserHeader?: string;
  EBlockUserDescription?: string;
  CauseBlockArray?: {
    id?: 'first' | 'second' | 'third' | 'fourth';
    title?: string;
    enabled?: boolean;
    reasonFor?: 'PLASTIC' | 'ALL';
  }[];
};

export type CreditCard = EntityByType<'CreditCard'> & { serviceList?: ServiceListMsa };
