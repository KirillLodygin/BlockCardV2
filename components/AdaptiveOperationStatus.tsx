import React from 'react';
import styled from 'styled-components/macro';

import { AdditionalTypography, ADDITIONAL_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/additionalTypography';
import { MainTypography, MAIN_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/mainTypography';
import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';
import { INDENT } from '@vtb/ui-kit/tokens/sizes';
import { TYPOGRAPHY_TAG } from '@vtb/ui-kit/tokens/typography';

import {
  OperationFormStatusAdaptive,
  StatusIconWrapperAdaptive,
  StatusInfoWrapper,
} from '@common/overmind/layouts/common/components/operationForm/AdaptiveOperationFormStatus';
import { ErrorAnimation, SuccessAnimation } from '@common/overmind/layouts/common/layoutParts/NotificationAnimation';

type NotificationType = typeof NOTIFICATION_TYPE.info | typeof NOTIFICATION_TYPE.fail;

type Props = {
  description: string;
  title: string;
  type: NotificationType;
};

const OperationFormStatusAdaptiveWrapper = styled(OperationFormStatusAdaptive)`
  margin-bottom: ${INDENT.m};
`;

const StatusTextAdaptive = styled(MainTypography)`
  margin-bottom: ${INDENT.xxxs};
`;

export const AdaptiveOperationStatus = ({ description, title, type }: Props) => (
  <OperationFormStatusAdaptiveWrapper isError={type === NOTIFICATION_TYPE.fail}>
    <StatusIconWrapperAdaptive>
      {type === NOTIFICATION_TYPE.fail ? <ErrorAnimation /> : <SuccessAnimation />}
    </StatusIconWrapperAdaptive>
    <StatusInfoWrapper>
      <StatusTextAdaptive tag={TYPOGRAPHY_TAG.h2} size={MAIN_TYPOGRAPHY_SIZE.m} autoFocus>
        {title}
      </StatusTextAdaptive>
      <AdditionalTypography size={ADDITIONAL_TYPOGRAPHY_SIZE.m}>{description}</AdditionalTypography>
    </StatusInfoWrapper>
  </OperationFormStatusAdaptiveWrapper>
);
