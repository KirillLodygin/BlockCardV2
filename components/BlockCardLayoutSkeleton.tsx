import React from 'react';
import { Flex, Box } from 'reflexbox/styled-components';

import { AdaptiveCol, AdaptiveRow } from '@vtb/ui-kit/adaptive/atoms/layout';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import {
  FieldSkeleton,
  DropdownSkeleton,
  TextlineSkeleton,
  LINE_SIZES,
  SKELETON_LINE_WIDTH,
} from '@components/Skeleton/components';
import { ButtonSkeleton } from '@components/Skeleton/components/adaptive/ButtonSkeleton';

const BlockCardLayoutSkeleton = () => (
  <Box mt={INDENT.m}>
    <AdaptiveRow>
      <AdaptiveCol tablet={8}>
        <Flex flexDirection="column">
          <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.l} />
          <Box mt={INDENT.sm}>
            <FieldSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.xxs}>
            <DropdownSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.l}>
            <FieldSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.xxs}>
            <ButtonSkeleton />
          </Box>
          <Box mt={INDENT.l}>
            <TextlineSkeleton width={SKELETON_LINE_WIDTH.xs} lineSize={LINE_SIZES.l} />
          </Box>
        </Flex>
      </AdaptiveCol>
    </AdaptiveRow>
  </Box>
);

export { BlockCardLayoutSkeleton };
