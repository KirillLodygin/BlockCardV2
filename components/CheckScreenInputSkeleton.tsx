import React, { FC } from 'react';
import { Box } from 'reflexbox/styled-components';

import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { FieldSkeleton, TextlineSkeleton, LINE_SIZES, SKELETON_LINE_WIDTH } from '@components/Skeleton/components';

export const CheckScreenInputSkeleton: FC = () => (
  <Box mb={INDENT.l}>
    <Box mb={INDENT.sm}>
      <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.m} />
    </Box>
    <Box maxWidth={FIELD_WIDTH.xl}>
      <FieldSkeleton width={FIELD_WIDTH.auto} />
    </Box>
    <Box mt={INDENT.l}>
      <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.m} />
    </Box>
  </Box>
);
