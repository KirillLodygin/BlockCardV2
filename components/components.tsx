import React, { useEffect, useRef, useState } from 'react';
import { Box, Flex } from 'reflexbox/styled-components';
import styled from 'styled-components/macro';

import { AdditionalTypography, ADDITIONAL_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/additionalTypography';
import { ButtonText } from '@vtb/ui-kit/adaptive/atoms/button';
import { Icon, ICON_SIZE } from '@vtb/ui-kit/adaptive/atoms/icon';
import { MAIN_TYPOGRAPHY_SIZE, MainTypography } from '@vtb/ui-kit/adaptive/atoms/mainTypography';
import { Notification as Notice, NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';
import { COLOR_GETTERS } from '@vtb/ui-kit/adaptive/tokens/palette/colorGetters';
import { CardPending, ChevronLeft } from '@vtb/ui-kit/assets';
import { INDENT } from '@vtb/ui-kit/tokens/sizes';
import { TYPOGRAPHY_WEIGHT } from '@vtb/ui-kit/tokens/typography';

import { ReactComponent as CardBlockErrorDark } from '@assets/cards/cardBlockErrorDark.svg';
import { ReactComponent as CardBlockErrorLight } from '@assets/cards/cardBlockErrorLight.svg';
import { ReactComponent as CardBlockSuccessDark } from '@assets/cards/cardBlockSuccessDark.svg';
import { ReactComponent as CardBlockSuccessLight } from '@assets/cards/cardBlockSuccessLight.svg';
import { useThemedContent } from '@common/appTheme';
import { OPERATIONS_CONSTANTS } from '@common/overmind/constants/base';
import { DEFAULT_ERROR_MSG } from '@common/overmind/constants/messages';
import { OperationFormChild } from '@common/overmind/layouts/common/components/operationForm';
import {
  OperationFormStatus,
  StatusIconContainer,
  StatusText,
} from '@common/overmind/layouts/common/components/operationForm/OperationFormStatus';
import { ELEMENT_TYPES } from '@common/overmind/layouts/common/constants';
import { cutBBCodesAndTags, parseBBCode } from '@common/parse-bb-code';
import { useSetNavBarTitle } from '@modules/NavBar';

// TODO-gordeev: использовать более подходящие типы после того как они будут сформированы в овермайнде
// TODO-svetlov: перетащить все компоненты в свои файлы и переместить в папку с компонентами

type TitleProps = {
  titleText?: string;
};

export const Title = ({ titleText }: TitleProps) => (
  <Box mb={INDENT.sm}>
    <MainTypography autoFocus>{parseBBCode(titleText)}</MainTypography>
  </Box>
);

const NotificationText = styled(AdditionalTypography).attrs({
  size: ADDITIONAL_TYPOGRAPHY_SIZE.m,
})`
  color: ${COLOR_GETTERS.text.secondary};
`;

type NotificationProps = {
  hasErrorNotification: boolean;
  isLockingSuccess: boolean;
  statusText: string | undefined;
  details: React.Component;
};

export const Notification = ({ statusText, hasErrorNotification, isLockingSuccess, details }: NotificationProps) => {
  const illustrations = useThemedContent(
    {
      success: <CardBlockSuccessLight />,
      error: <CardBlockErrorLight />,
    },
    {
      success: <CardBlockSuccessDark />,
      error: <CardBlockErrorDark />,
    },
  );

  return (
    <Box mb={INDENT.l}>
      <OperationFormStatus>
        <StatusIconContainer>
          {isLockingSuccess && illustrations.success}
          {hasErrorNotification && illustrations.error}
        </StatusIconContainer>
        <StatusText autoFocus>{statusText}</StatusText>

        <NotificationText>{details}</NotificationText>
      </OperationFormStatus>
    </Box>
  );
};

type ButtonsBlockProps = {
  buttons: ReadonlyArray<any>;
};

export const ButtonsBlock = ({ buttons }: ButtonsBlockProps) => (
  <Box mt={INDENT.m}>
    {buttons.map((button) =>
      button.props.name === OPERATIONS_CONSTANTS.commandNames.CANCEL ? (
        <Box mt={INDENT.l} key={button.key}>
          {button}
        </Box>
      ) : (
        <Box key={button.key}>{button}</Box>
      ),
    )}
  </Box>
);

type ElementsBlockProps = {
  elements: ReadonlyArray<any>;
};

export const ElementsBlock = ({ elements }: ElementsBlockProps) => (
  <Box mb={INDENT.xs} flexDirection="column" key="rootGroup">
    {elements.map((field, i) => (
      <OperationFormChild type={ELEMENT_TYPES.FIELD} key={i}>
        {field}
      </OperationFormChild>
    ))}
  </Box>
);

export const BlockTitle = ({ titleText }: TitleProps) => {
  const TitleRef = useRef(null);
  useSetNavBarTitle(titleText, TitleRef);

  if (titleText) {
    return (
      <Box mb={INDENT.m}>
        <MainTypography size={MAIN_TYPOGRAPHY_SIZE.m} tag="h3" ref={TitleRef} autoFocus>
          {titleText}
        </MainTypography>
      </Box>
    );
  }

  return null;
};

export const DefaultErrorBlock = () => (
  <Box mt={INDENT.m}>
    <Notice text={DEFAULT_ERROR_MSG} type={NOTIFICATION_TYPE.fail} />
  </Box>
);

const StyledAdditionalTypography = styled(AdditionalTypography)`
  color: ${COLOR_GETTERS.text.primary};
`;

type CodeRequestTimerProps = {
  isTimerRunning: boolean;
  setIsTimerRunning: (value: boolean) => void;
  text: string;
  timeout: number;
};

export const CodeRequestTimer = ({ isTimerRunning, setIsTimerRunning, text, timeout }: CodeRequestTimerProps) => {
  const [seconds, setSeconds] = useState(timeout);

  useEffect(() => {
    const timerId = setInterval(() => {
      setSeconds(seconds - 1);
    }, 1000);

    if (seconds === 0) {
      clearInterval(timerId);
      setIsTimerRunning(false);
    }

    return () => clearInterval(timerId);
  }, [seconds, setIsTimerRunning]);

  useEffect(() => {
    if (isTimerRunning) {
      setSeconds(timeout);
    }
  }, [isTimerRunning, timeout]);

  return (
    <Box mt={INDENT.xs}>
      <Flex>
        {isTimerRunning && <Icon type={CardPending} size={ICON_SIZE.s} />}
        <StyledAdditionalTypography weight={TYPOGRAPHY_WEIGHT.bold} role="timer">
          <Box as="span" ml={INDENT.xxxxs}>
            {cutBBCodesAndTags(text, seconds.toString().padStart(2, '0'))}
          </Box>
        </StyledAdditionalTypography>
      </Flex>
    </Box>
  );
};

type BackButtonBlockProps = {
  children: string;
  'data-test-id': string;
  disabled?: boolean;
  onClick: () => void;
};

export const BackButtonBlock = ({ children, ...props }: BackButtonBlockProps) => (
  <ButtonText size="medium" icon={ChevronLeft} {...props}>
    {children}
  </ButtonText>
);
