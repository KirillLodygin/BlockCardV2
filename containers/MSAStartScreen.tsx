import React, { useCallback } from 'react';
import { Box } from 'reflexbox/styled-components';

import { getDataTestId } from '@vtb/ui-kit/_utils/getDataTestId';
import { ButtonContained } from '@vtb/ui-kit/adaptive/atoms/button';
import { Input } from '@vtb/ui-kit/adaptive/atoms/input';
import { Radio } from '@vtb/ui-kit/adaptive/atoms/radio';
import { Dropdown } from '@vtb/ui-kit/adaptive/molecules/dropdown';
import { DropdownValue } from '@vtb/ui-kit/adaptive/molecules/dropdown/_baseDropdown/typesDropdown';
import { Notification } from '@vtb/ui-kit/adaptive/molecules/notification';
import { RADIO_GROUP_TYPE, RadioGroup } from '@vtb/ui-kit/adaptive/molecules/radioGroup';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { BackButtonBlock, BlockTitle } from '../components/components';
import { LOCK_OPERATION, LOCK_REASON, BUTTONS_LOCK, NOTIFICATION_TYPE } from '../types';
import { getNotificationType } from '../utils/notificationUtils';

type Props = {
  content: {
    buttons: {
      onClick: () => void;
      title: string;
      type: BUTTONS_LOCK;
    }[];
    buttonsRadio: {
      id: LOCK_OPERATION;
      text: string;
    }[];
    dropdownLabel: string;
    dropdownOptions: {
      id: LOCK_REASON;
      caption: string;
      value: string;
    }[];
    inputLabel: string;
    notification: {
      description: string;
      type: NOTIFICATION_TYPE;
    };
    title: string;
  };
  inputValue: string;
  isFetchLoading: boolean;
  isTempBlockAvailable: boolean;
  onChangeBlockCardType: (selectedRadioButton: LOCK_OPERATION.TEMPORARY | LOCK_OPERATION.PERMANENT) => void;
  onChangeReasonBlockCard: (reasonBlockCard: DropdownValue) => void;
  reasonBlockCard?: DropdownValue;
  selectedRadioButton: LOCK_OPERATION;
};

export const MSAStartScreen = ({
  content: {
    buttons,
    buttonsRadio,
    dropdownLabel,
    dropdownOptions,
    inputLabel,
    notification: { description: notificationDescription, type: notificationType },
    title,
  },
  inputValue,
  isFetchLoading,
  isTempBlockAvailable,
  onChangeBlockCardType,
  onChangeReasonBlockCard,
  reasonBlockCard,
  selectedRadioButton,
}: Props) => {
  const handleChangeRadioButton = useCallback(
    (_: boolean, value: LOCK_OPERATION.TEMPORARY | LOCK_OPERATION.PERMANENT) => {
      onChangeBlockCardType(value);
    },
    [onChangeBlockCardType],
  );

  return (
    <>
      <BlockTitle titleText={title} />
      {isTempBlockAvailable && (
        <Box mb={INDENT.xs}>
          <RadioGroup
            data-test-id={getDataTestId('RadioGroup', 'blockCard')}
            disabled={isFetchLoading}
            type={RADIO_GROUP_TYPE.horizontal}
            onChange={handleChangeRadioButton}
            selected={selectedRadioButton}
            name="RadioGroup"
          >
            {buttonsRadio.map(({ id, text }) => (
              <Radio key={id} value={`${id}`}>
                {text}
              </Radio>
            ))}
          </RadioGroup>
        </Box>
      )}
      <Box mb={INDENT.l} maxWidth={FIELD_WIDTH.xl}>
        <Input label={inputLabel} value={inputValue} disabled />
      </Box>
      {selectedRadioButton === LOCK_OPERATION.PERMANENT && (
        <Box mb={INDENT.l} maxWidth={FIELD_WIDTH.xl}>
          <Dropdown
            data-test-id="whyBlockingCardDropdown"
            label={dropdownLabel}
            options={dropdownOptions}
            onChange={onChangeReasonBlockCard}
            value={reasonBlockCard}
          />
        </Box>
      )}
      <Box mb={INDENT.m} maxWidth={FIELD_WIDTH.xl}>
        <Notification type={getNotificationType(notificationType)} text={notificationDescription} />
      </Box>
      {buttons.map(({ onClick, title, type }, index) => {
        const mt = index > 0 ? INDENT.m : undefined;
        const ButtonComponent = type === BUTTONS_LOCK.CONTINUE ? ButtonContained.Loader : BackButtonBlock;
        const isLoading = type === BUTTONS_LOCK.CONTINUE && isFetchLoading;

        return (
          <Box mt={mt} key={`${title}_${type}`}>
            <ButtonComponent
              data-test-id={getDataTestId(type, 'blockCardButton')}
              onClick={onClick}
              isLoading={isLoading}
              disabled={isFetchLoading}
            >
              {title}
            </ButtonComponent>
          </Box>
        );
      })}
    </>
  );
};
