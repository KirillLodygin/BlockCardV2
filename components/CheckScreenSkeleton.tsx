import React from 'react';
import { Flex, Box } from 'reflexbox/styled-components';

import { AdaptiveCol, AdaptiveRow } from '@vtb/ui-kit/adaptive/atoms/layout';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { LINE_SIZES, SKELETON_LINE_WIDTH, TextlineSkeleton } from '@components/Skeleton/components';
import { ButtonSkeleton } from '@components/Skeleton/components/adaptive/ButtonSkeleton';

import { CheckScreenInputSkeleton } from './CheckScreenInputSkeleton';
import { DataRowTableSkeleton } from './DataRowTableSkeleton';

const CheckScreenSkeleton = () => (
  <Box mt={INDENT.m}>
    <AdaptiveRow>
      <AdaptiveCol tablet={8}>
        <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.l} />
        <DataRowTableSkeleton />
        <CheckScreenInputSkeleton />
        <Flex alignItems="center" flexDirection="row" maxWidth={FIELD_WIDTH.xl} mb={INDENT.m}>
          <ButtonSkeleton buttonBorderRadius={4} />
          <Box ml={INDENT.m}>
            <ButtonSkeleton buttonBorderRadius={4} />
          </Box>
        </Flex>
        <Box mt={INDENT.l}>
          <TextlineSkeleton width={SKELETON_LINE_WIDTH.xs} lineSize={LINE_SIZES.l} />
        </Box>
      </AdaptiveCol>
    </AdaptiveRow>
  </Box>
);

export { CheckScreenSkeleton };
