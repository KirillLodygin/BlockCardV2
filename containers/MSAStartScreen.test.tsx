import { shallow } from 'enzyme';
import React from 'react';

import { ButtonContained } from '@vtb/ui-kit/adaptive/atoms/button';
import { Dropdown } from '@vtb/ui-kit/adaptive/molecules/dropdown';
import { RadioGroup } from '@vtb/ui-kit/adaptive/molecules/radioGroup';

import { BackButtonBlock } from '../components/components';
import { LOCK_OPERATION, BUTTONS_LOCK, LOCK_REASON, NOTIFICATION_TYPE } from '../types';
import { MSAStartScreen } from './MSAStartScreen';

const mockOnChangeBlockCardType = jest.fn();
const mockOnChangeReasonBlockCard = jest.fn();
const mockOnConfirm = jest.fn();
const mockOnBack = jest.fn();

const reasonBlockCard = {
  id: LOCK_REASON.LOSS,
  caption: 'foo3',
  value: 'foo4',
};

const dropdownOptions = [
  {
    id: LOCK_REASON.THEFT,
    caption: 'foo1',
    value: 'foo2',
  },
  reasonBlockCard,
];

const buttonsRadio = [
  {
    id: LOCK_OPERATION.PERMANENT,
    text: 'baz1',
  },
  {
    id: LOCK_OPERATION.TEMPORARY,
    text: 'baz2',
  },
];

const defaultProps = {
  content: {
    buttons: [
      {
        onClick: mockOnConfirm,
        title: 'asd',
        type: BUTTONS_LOCK.CONTINUE,
      },
      {
        onClick: mockOnBack,
        title: 'qwe',
        type: BUTTONS_LOCK.BACK,
      },
    ],
    dropdownOptions,
    buttonsRadio,
    cardTitle: 'bat',
    dropdownLabel: 'bar',
    inputLabel: 'foo',
    notification: {
      description: 'sdf',
      type: NOTIFICATION_TYPE.INFO,
    },
    title: 'wer',
  },
  inputValue: 'qux',
  isFetchLoading: false,
  isTempBlockAvailable: true,
  onChangeBlockCardType: mockOnChangeBlockCardType,
  onChangeReasonBlockCard: mockOnChangeReasonBlockCard,
  reasonBlockCard,
  selectedRadioButton: LOCK_OPERATION.PERMANENT,
};

jest.mock('../utils/notificationUtils', () => ({
  getNotificationType: jest.fn().mockReturnValue('info'),
}));

describe('MSAStartScreen', () => {
  beforeEach(() => {
    mockOnBack.mockReset();
    mockOnConfirm.mockReset();
    mockOnChangeBlockCardType.mockReset();
    mockOnChangeReasonBlockCard.mockReset();
  });

  it('renders with all props', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  it('renders with selected radio button = temporary', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} selectedRadioButton={LOCK_OPERATION.TEMPORARY} />);
    expect(component).toMatchSnapshot();
  });

  it('calls onChangeBlockCardType', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} />);

    component.find(RadioGroup).prop('onChange')?.(true, LOCK_OPERATION.TEMPORARY);

    expect(mockOnChangeBlockCardType).toHaveBeenCalledTimes(1);
    expect(mockOnChangeBlockCardType).toHaveBeenCalledWith(LOCK_OPERATION.TEMPORARY);
  });

  it('calls onChangeReasonBlockCard', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} />);

    component.find(Dropdown).simulate('change');

    expect(mockOnChangeReasonBlockCard).toHaveBeenCalledTimes(1);
  });

  it('calls onCanfirm', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} />);

    component.find(ButtonContained.Loader).simulate('click');

    expect(mockOnConfirm).toHaveBeenCalledTimes(1);
  });

  it('calls onBack', () => {
    const component = shallow(<MSAStartScreen {...defaultProps} />);

    component.find(BackButtonBlock).simulate('click');

    expect(mockOnBack).toHaveBeenCalledTimes(1);
  });
});
