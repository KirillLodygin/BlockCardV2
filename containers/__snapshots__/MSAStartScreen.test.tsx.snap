// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MSAStartScreen renders with all props 1`] = `
<Fragment>
  <BlockTitle
    titleText="wer"
  />
  <styled.div
    mb="16px"
  >
    <RadioGroup
      data-test-id="radiogroup_blockcard"
      disabled={false}
      name="RadioGroup"
      onChange={[Function]}
      selected="permanent"
      type="horizontal"
    >
      <Radio
        key="permanent"
        value="permanent"
      >
        baz1
      </Radio>
      <Radio
        key="temporary"
        value="temporary"
      >
        baz2
      </Radio>
    </RadioGroup>
  </styled.div>
  <styled.div
    maxWidth="544px"
    mb="40px"
  >
    <Input
      disabled={true}
      label="foo"
      value="qux"
    />
  </styled.div>
  <styled.div
    maxWidth="544px"
    mb="40px"
  >
    <Dropdown
      data-test-id="whyBlockingCardDropdown"
      label="bar"
      onChange={[MockFunction]}
      options={
        Array [
          Object {
            "caption": "foo1",
            "id": "theft",
            "value": "foo2",
          },
          Object {
            "caption": "foo3",
            "id": "loss",
            "value": "foo4",
          },
        ]
      }
      value={
        Object {
          "caption": "foo3",
          "id": "loss",
          "value": "foo4",
        }
      }
    />
  </styled.div>
  <styled.div
    maxWidth="544px"
    mb="32px"
  >
    <Notification
      text="sdf"
      type="info"
    />
  </styled.div>
  <styled.div
    key="asd_continue"
  >
    <ButtonContainedLoader
      data-test-id="continue_blockcardbutton"
      disabled={false}
      isLoading={false}
      onClick={[MockFunction]}
    >
      asd
    </ButtonContainedLoader>
  </styled.div>
  <styled.div
    key="qwe_back"
    mt="32px"
  >
    <BackButtonBlock
      data-test-id="back_blockcardbutton"
      disabled={false}
      isLoading={false}
      onClick={[MockFunction]}
    >
      qwe
    </BackButtonBlock>
  </styled.div>
</Fragment>
`;

exports[`MSAStartScreen renders with selected radio button = temporary 1`] = `
<Fragment>
  <BlockTitle
    titleText="wer"
  />
  <styled.div
    mb="16px"
  >
    <RadioGroup
      data-test-id="radiogroup_blockcard"
      disabled={false}
      name="RadioGroup"
      onChange={[Function]}
      selected="temporary"
      type="horizontal"
    >
      <Radio
        key="permanent"
        value="permanent"
      >
        baz1
      </Radio>
      <Radio
        key="temporary"
        value="temporary"
      >
        baz2
      </Radio>
    </RadioGroup>
  </styled.div>
  <styled.div
    maxWidth="544px"
    mb="40px"
  >
    <Input
      disabled={true}
      label="foo"
      value="qux"
    />
  </styled.div>
  <styled.div
    maxWidth="544px"
    mb="32px"
  >
    <Notification
      text="sdf"
      type="info"
    />
  </styled.div>
  <styled.div
    key="asd_continue"
  >
    <ButtonContainedLoader
      data-test-id="continue_blockcardbutton"
      disabled={false}
      isLoading={false}
      onClick={[MockFunction]}
    >
      asd
    </ButtonContainedLoader>
  </styled.div>
  <styled.div
    key="qwe_back"
    mt="32px"
  >
    <BackButtonBlock
      data-test-id="back_blockcardbutton"
      disabled={false}
      isLoading={false}
      onClick={[MockFunction]}
    >
      qwe
    </BackButtonBlock>
  </styled.div>
</Fragment>
`;
